File: day-11/python/main.py
import pathlib
import itertools
from copy import deepcopy
from typing import Callable


OCCUPIED_SEAT = '#'
EMPTY_SEAT = 'L'
FLOOR = '.'


def get_occuppied_seat_total(seat_layout: list[list[str]]) -> int:
    return sum(row.count(OCCUPIED_SEAT) for row in seat_layout)


def occupied_adjacent_seats(x: int, y: int, seat_layout: list[list[str]]) -> int:
    return sum(
        seat_layout[y][x] == OCCUPIED_SEAT
        for x, y in {
            (x + i, y + j)
            for i in range(-1, 2)
            for j in range(-1, 2)
            if not (i == 0 and j == 0)
        }
        if x in range(len(seat_layout[0])) and y in range(len(seat_layout))
    )


def occupied_visible_seats(x: int, y: int, seat_layout: list[list[str]]) -> int:
    rows = range(len(seat_layout))
    columns = range(len(seat_layout[0]))
    visibile_direction_deltas = {
        d for d in itertools.product(range(-1, 2), range(-1, 2)) if d != (0, 0)
    }

    occupied_visible_seat_count = 0
    for x_delta, y_delta in visibile_direction_deltas:
        visible_x, visible_y = x + x_delta, y + y_delta
        while visible_x in columns and visible_y in rows:
            visible_seat = seat_layout[visible_y][visible_x]
            if visible_seat == FLOOR:
                visible_x += x_delta
                visible_y += y_delta
                continue
            if visible_seat == OCCUPIED_SEAT:
                occupied_visible_seat_count += 1
            break

    return occupied_visible_seat_count


def execute_step(
    seat_layout: list[list[str]], occupancy_criteria: Callable, occupancy_threshold: int
) -> list[list[str]]:
    pending_operations = []
    for y in range(len(seat_layout)):
        for x in range(len(seat_layout[0])):
            this_seat = seat_layout[y][x]
            occupied_seat_count = occupancy_criteria(x, y, seat_layout)
            if this_seat == EMPTY_SEAT and occupied_seat_count == 0:
                pending_operations.append(((x, y), OCCUPIED_SEAT))
            elif (
                this_seat == OCCUPIED_SEAT
                and occupied_seat_count >= occupancy_threshold
            ):
                pending_operations.append(((x, y), EMPTY_SEAT))
    for indices, value in pending_operations:
        seat_layout[indices[1]][indices[0]] = value
    return seat_layout


def execute(
    seat_layout: list[list[str]], occupancy_criteria: Callable, occupancy_threshold: int
) -> int:
    while seat_layout != (
        new_seat_layout := execute_step(
            deepcopy(seat_layout), occupancy_criteria, occupancy_threshold
        )
    ):
        seat_layout = new_seat_layout
    return get_occuppied_seat_total(seat_layout)


with open(pathlib.Path(__file__).parent.parent / 'input.txt') as f:
    seat_layout = [list(line.strip()) for line in f.read().splitlines()]

part_one_solution = execute(seat_layout, occupied_adjacent_seats, 4)
part_two_solution = execute(seat_layout, occupied_visible_seats, 5)

print('Part One:', part_one_solution)
print('Part Two:', part_two_solution)

assert part_one_solution == 2453
assert part_two_solution == 2159

File: day-11/rust/src/main.rs
#[macro_use]
extern crate lazy_static;

use std::fs;
use std::path::Path;

use itertools::Itertools;

const OCCUPIED_SEAT: char = '#';
const EMPTY_SEAT: char = 'L';
const FLOOR: char = '.';

lazy_static! {
    static ref DIRECTION_UNIT_DELTAS: Vec<(i32, i32)> = (-1..2)
        .cartesian_product(-1..2)
        .filter(|(x, y)| !(*x == 0 && *y == 0))
        .collect();
}

fn get_occupied_seat_total(seat_layout: Vec<Vec<char>>) -> usize {
    seat_layout
        .iter()
        .map(|row| row.iter().filter(|seat| **seat == OCCUPIED_SEAT).count())
        .sum()
}

fn occupied_adjacent_seats(x: usize, y: usize, seat_layout: &[Vec<char>]) -> usize {
    DIRECTION_UNIT_DELTAS
        .iter()
        .map(|(x_delta, y_delta)| (x as i32 + x_delta, y as i32 + y_delta))
        .filter(|(adjacent_x, adjacent_y)| {
            let (adjacent_x, adjacent_y) = (*adjacent_x as usize, *adjacent_y as usize);
            (0..seat_layout.len()).contains(&(adjacent_y))
                && (0..seat_layout[0].len()).contains(&(adjacent_x))
                && seat_layout[adjacent_y][adjacent_x] == OCCUPIED_SEAT
        })
        .count()
}

fn occupied_visible_seats(x: usize, y: usize, seat_layout: &[Vec<char>]) -> usize {
    let rows = seat_layout.len();
    let columns = seat_layout[0].len();

    let mut occupied_visible_seat_count = 0;

    for (x_delta, y_delta) in DIRECTION_UNIT_DELTAS.iter() {
        let (mut visible_x, mut visible_y) = (x as i32 + x_delta, y as i32 + y_delta);
        while (0..columns).contains(&(visible_x as usize))
            && (0..rows).contains(&(visible_y as usize))
        {
            match seat_layout[visible_y as usize][visible_x as usize] {
                FLOOR => {
                    visible_x += x_delta;
                    visible_y += y_delta;
                    continue;
                }
                OCCUPIED_SEAT => occupied_visible_seat_count += 1,
                _ => (),
            }
            break;
        }
    }
    occupied_visible_seat_count
}

fn execute_step(
    mut seat_layout: Vec<Vec<char>>,
    occupancy_criteria: fn(usize, usize, &[Vec<char>]) -> usize,
    occupancy_threshold: usize,
) -> Vec<Vec<char>> {
    let mut pending_operations: Vec<((usize, usize), char)> = vec![];
    for y in 0..seat_layout.len() {
        for x in 0..seat_layout[0].len() {
            let this_seat = seat_layout[y][x];
            let occupied_seat_count = occupancy_criteria(x, y, &seat_layout);
            match this_seat {
                EMPTY_SEAT => {
                    if occupied_seat_count == 0 {
                        pending_operations.push(((x, y), OCCUPIED_SEAT));
                    }
                }
                OCCUPIED_SEAT => {
                    if occupied_seat_count >= occupancy_threshold {
                        pending_operations.push(((x, y), EMPTY_SEAT));
                    }
                }
                _ => (),
            }
        }
    }

    for (indices, value) in pending_operations.iter() {
        seat_layout[indices.1][indices.0] = *value;
    }

    seat_layout
}

fn execute(
    seat_layout: &[Vec<char>],
    occupancy_criteria: fn(usize, usize, &[Vec<char>]) -> usize,
    occupancy_threshold: usize,
) -> usize {
    let (mut old_seat_layout, mut new_seat_layout) = (vec![], seat_layout.to_vec());
    while !old_seat_layout.iter().eq(new_seat_layout.iter()) {
        old_seat_layout = new_seat_layout;
        new_seat_layout = execute_step(
            old_seat_layout.clone(),
            occupancy_criteria,
            occupancy_threshold,
        );
    }
    get_occupied_seat_total(new_seat_layout)
}

fn main() {
    let seat_layout: Vec<Vec<_>> = fs::read_to_string(
        Path::new(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .join("input.txt"),
    )
    .unwrap()
    .lines()
    .map(|l| l.parse::<String>().unwrap().chars().collect())
    .collect();

    let part_one_solution = execute(seat_layout.as_slice(), occupied_adjacent_seats, 4);
    let part_two_solution = execute(seat_layout.as_slice(), occupied_visible_seats, 5);

    println!("Part One: {}", &part_one_solution);
    println!("Part Two: {}", &part_two_solution);

    assert_eq!(part_one_solution, 2453);
    assert_eq!(part_two_solution, 2159);
}
